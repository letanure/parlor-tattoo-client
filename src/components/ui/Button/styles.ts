import styled, { css, DefaultTheme } from 'styled-components'

import { ColorTypes } from 'styles/theme'
import { ButtonProps } from '.'

type WrapperProps = ButtonProps

const wrapperModifiers = {
  outline: (theme: DefaultTheme, action: ColorTypes) => css`
    background-color: transparent;
    border: 1px solid ${theme.color[action!].normal.value};
    color: ${theme.color[action!].normal.value};
    &:hover {
      background-color: ${theme.color[action!].hover.value};
      border-color: ${theme.color[action!].hover.value};
      color: ${theme.color[action!].hover.inverse};
    }
  `,
  disabled: (theme: DefaultTheme, action: ColorTypes) => css`
    background-color: ${theme.color[action!].disabled.value};
    color: ${theme.color[action!].disabled.inverse};
    cursor: not-allowed;
  `,
  fullWidth: css`
    width: 100%;
  `
}

export const Wrapper = styled.button<WrapperProps>`
  ${({ action, disabled, fullWidth, outline, size, theme }) => css`
    background-color: ${theme.color[action!].normal.value};
    border-radius: 0.5ch;
    border: 1px solid ${theme.color[action!].normal.value};
    color: ${theme.color[action!].normal.inverse};
    cursor: pointer;
    font-size: ${theme.font.size[size!]};
    padding: 1ch 2ch;
    transition: all 0.1s ease-in-out;
    &:hover {
      background-color: ${theme.color[action!].hover.value};
      border-color: ${theme.color[action!].hover.value};
      color: ${theme.color[action!].hover.inverse};
    }

    ${outline && wrapperModifiers.outline(theme, action!)}
    ${disabled && wrapperModifiers.disabled(theme, action!)}
    ${fullWidth && wrapperModifiers.fullWidth}
  `}
`
