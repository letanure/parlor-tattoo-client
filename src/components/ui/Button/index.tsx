import { SizeOptions, ColorTypes } from 'styles/theme'
import * as S from './styles'

export type ButtonProps = {
  action?: ColorTypes
  children?: React.ReactNode
  disabled?: boolean
  outline?: boolean
  size?: SizeOptions
}

const Button = ({
  action = 'primary',
  children,
  disabled = false,
  outline = false,
  size = 'medium'
}: ButtonProps) => (
  <S.Wrapper action={action} disabled={disabled} outline={outline} size={size}>
    {!!children && <span>{children}</span>}
  </S.Wrapper>
)

export default Button
