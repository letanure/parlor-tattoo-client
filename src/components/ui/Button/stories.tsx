import { Story, Meta } from '@storybook/react/types-6-0'
import { Lock } from '@styled-icons/octicons'

import Button, { ButtonProps } from '.'

export default {
  title: 'UI/Button',
  component: Button,
  args: {
    children: 'My Button',
    disabled: false,
    outline: false,
    fullWidth: false
  }
} as Meta

export const Default: Story<ButtonProps> = (args) => <Button {...args} />

export const withIcon: Story<ButtonProps> = (args) => <Button {...args} />

withIcon.args = {
  icon: <Lock />
}
