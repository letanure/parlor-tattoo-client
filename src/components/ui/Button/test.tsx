import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/tests/helpers'

import Button from '.'

describe('<Button />', () => {
  it('should render a a text inside', () => {
    renderWithTheme(<Button>Text</Button>)
    expect(screen.getByRole('button')).toHaveTextContent('Text')
  })

  it('should render a primary button by default', () => {
    renderWithTheme(<Button />)
    expect(screen.getByRole('button')).toHaveStyle({
      background: '#8943e3'
    })
  })

  it('should render a secondary if provided', () => {
    renderWithTheme(<Button action="secondary" />)
    expect(screen.getByRole('button')).toHaveStyle({
      background: '#26a69a'
    })
  })

  it('should render a bigger button if provided', () => {
    renderWithTheme(<Button size="xxlarge" />)
    expect(screen.getByRole('button')).toHaveStyle({
      fontSize: '2.8rem'
    })
  })

  it('should render a outline version if provided', () => {
    renderWithTheme(<Button outline />)
    expect(screen.getByRole('button')).toHaveStyle({
      background: 'transparent'
    })
  })

  it('should render a disabled version if provided', () => {
    renderWithTheme(<Button disabled />)
    expect(screen.getByRole('button')).toHaveStyle({
      cursor: 'not-allowed'
    })
  })
})
