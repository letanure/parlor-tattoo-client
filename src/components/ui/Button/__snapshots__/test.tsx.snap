// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render a a text inside 1`] = `
.c0 {
  background-color: #8943e3;
  border-radius: 0.5ch;
  border: 1px solid #8943e3;
  color: #f5f5f5;
  cursor: pointer;
  font-size: 1.6rem;
  padding: 1ch 2ch;
  -webkit-transition: all 0.1s ease-in-out;
  transition: all 0.1s ease-in-out;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0:hover {
  background-color: #a166f0;
  border-color: #a166f0;
  color: #212121;
}

.c0 span {
  white-space: nowrap;
}

<button
  action="primary"
  class="c0"
>
  <span>
    Text
  </span>
</button>
`
